File: Sistema_Buses_Proyecto/src/Logica/Opt_Rutas.java
package Logica;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Opt_Rutas {

    public void mostrarResultados() {
        String archivo = "C:\\Users\\User\\Desktop\\Talleres\\por ahora\\aab2-proyecto-grupo02\\Sistema_Buses_Proyecto\\src\\Datos\\Lineasbu - copia.csv";
        List<List<String>> lineasYParadas = leerArchivoCSV(archivo);

        if (lineasYParadas.isEmpty()) {
            System.out.println("No se encontraron datos en el archivo CSV.");
            return;
        }

        List<String> rutaOptimizada = optimizarRuta(lineasYParadas);
        mostrarResultados(rutaOptimizada, lineasYParadas);
    }

    public List<List<String>> leerArchivoCSV(String archivo) {
        List<List<String>> lineasYParadas = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(archivo))) {
            String linea;
            while ((linea = br.readLine()) != null) {
                String[] partes = linea.split(";");
                List<String> fila = new ArrayList<>();
                for (String parte : partes) {
                    fila.add(parte.trim());
                }
                lineasYParadas.add(fila);
            }
        } catch (IOException e) {
            System.err.println("Error al leer el archivo CSV: " + e.getMessage());
        }
        return lineasYParadas;
    }

    public List<String> optimizarRuta(List<List<String>> lineasYParadas) {
        List<String> rutaOptimizada = new ArrayList<>();
        Set<String> paradasVisitadas = new HashSet<>();

        if (lineasYParadas.isEmpty() || lineasYParadas.get(0).size() < 2) {
            System.out.println("Error: No hay suficientes datos para optimizar la ruta.");
            return rutaOptimizada;
        }

        String paradaActual = lineasYParadas.get(0).get(1); // Primera parada de la primera línea
        rutaOptimizada.add(paradaActual);
        paradasVisitadas.add(paradaActual);

        while (paradasVisitadas.size() < lineasYParadas.size()) {
            String siguienteParada = null;
            double distanciaMinima = Double.MAX_VALUE;

            for (List<String> linea : lineasYParadas) {
                if (linea.size() < 2) continue;
                String parada = linea.get(1);

                if (!paradasVisitadas.contains(parada)) {
                    double distancia = calcularDistancia(lineasYParadas, paradaActual, parada);
                    if (distancia < distanciaMinima) {
                        distanciaMinima = distancia;
                        siguienteParada = parada;
                    }
                }
            }

            if (siguienteParada != null) {
                rutaOptimizada.add(siguienteParada);
                paradasVisitadas.add(siguienteParada);
                paradaActual = siguienteParada;
            } else {
                break;
            }
        }

        return rutaOptimizada;
    }

    public double calcularDistancia(List<List<String>> lineasYParadas, String parada1, String parada2) {
        double radioTierra = 6371;

        List<String> datosParada1 = buscarParada(lineasYParadas, parada1);
        List<String> datosParada2 = buscarParada(lineasYParadas, parada2);

        if (datosParada1 == null || datosParada2 == null || datosParada1.size() < 4 || datosParada2.size() < 4) {
            return Double.MAX_VALUE;
        }

        try {
            double lat1 = Double.parseDouble(datosParada1.get(2));
            double lon1 = Double.parseDouble(datosParada1.get(3));
            double lat2 = Double.parseDouble(datosParada2.get(2));
            double lon2 = Double.parseDouble(datosParada2.get(3));

            double dLat = Math.toRadians(lat2 - lat1);
            double dLon = Math.toRadians(lon2 - lon1);
            double a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
                       Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
                       Math.sin(dLon / 2) * Math.sin(dLon / 2);
            double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
            return radioTierra * c;
        } catch (NumberFormatException e) {
            System.err.println("Error: Las coordenadas no son números válidos.");
            return Double.MAX_VALUE;
        }
    }

    public List<String> buscarParada(List<List<String>> lineasYParadas, String nombreParada) {
        for (List<String> linea : lineasYParadas) {
            if (linea.size() > 1 && linea.get(1).equals(nombreParada)) {
                return linea;
            }
        }
        return null;
    }

    public void mostrarResultados(List<String> rutaOptimizada, List<List<String>> lineasYParadas) {
        if (rutaOptimizada.isEmpty()) {
            System.out.println("No se encontraron datos para mostrar.");
            return;
        }

        System.out.println("=== RUTA OPTIMIZADA ===");
        System.out.println("+---------------------+----------------------+----------------------+");
        System.out.println("| Nombre de Parada    | Latitud              | Longitud             |");
        System.out.println("+---------------------+----------------------+----------------------+");

        for (String parada : rutaOptimizada) {
            List<String> datosParada = buscarParada(lineasYParadas, parada);
            if (datosParada != null && datosParada.size() > 3) {
                System.out.format("| %-20s | %-20s | %-20s |\n",
                        datosParada.get(1), datosParada.get(2), datosParada.get(3));
            }
        }

        System.out.println("+---------------------+----------------------+----------------------+");
    }
}
/**
 * 
 */

File: Python/Logica/optRutas.py
import math

class Opt_Rutas:
    def mostrar_resultados(self):
        archivo = "C:\Users\User\Desktop\Talleres\por ahora\aab2-proyecto-grupo02\Python\Datos\Lineasbu - copia.csv"
        lineas_y_paradas = self.leer_archivo_csv(archivo)

        if not lineas_y_paradas:
            print("No se encontraron datos en el archivo CSV.")
            return

        ruta_optimizada = self.optimizar_ruta(lineas_y_paradas)
        self.mostrar_resultados_finales(ruta_optimizada, lineas_y_paradas)

    def leer_archivo_csv(self, archivo):
        lineas_y_paradas = []
        try:
            with open(archivo, "r", encoding="utf-8") as archivo_csv:
                for linea in archivo_csv:
                    partes = linea.strip().split(";")
                    fila = [parte.strip() for parte in partes]
                    lineas_y_paradas.append(fila)
        except IOError as e:
            print(f"Error al leer el archivo CSV: {e}")
        return lineas_y_paradas

    def optimizar_ruta(self, lineas_y_paradas):
        ruta_optimizada = []
        paradas_visitadas = set()

        if not lineas_y_paradas or len(lineas_y_paradas[0]) < 2:
            print("Error: No hay suficientes datos para optimizar la ruta.")
            return ruta_optimizada

        parada_actual = lineas_y_paradas[0][1]  # Primera parada de la primera línea
        ruta_optimizada.append(parada_actual)
        paradas_visitadas.add(parada_actual)

        while len(paradas_visitadas) < len(lineas_y_paradas):
            siguiente_parada = None
            distancia_minima = float("inf")

            for linea in lineas_y_paradas:
                if len(linea) < 2:
                    continue
                parada = linea[1]

                if parada not in paradas_visitadas:
                    distancia = self.calcular_distancia(lineas_y_paradas, parada_actual, parada)
                    if distancia < distancia_minima:
                        distancia_minima = distancia
                        siguiente_parada = parada

            if siguiente_parada:
                ruta_optimizada.append(siguiente_parada)
                paradas_visitadas.add(siguiente_parada)
                parada_actual = siguiente_parada
            else:
                break

        return ruta_optimizada

    def calcular_distancia(self, lineas_y_paradas, parada1, parada2):
        radio_tierra = 6371

        datos_parada1 = self.buscar_parada(lineas_y_paradas, parada1)
        datos_parada2 = self.buscar_parada(lineas_y_paradas, parada2)

        if not datos_parada1 or not datos_parada2 or len(datos_parada1) < 4 or len(datos_parada2) < 4:
            return float("inf")

        try:
            lat1 = float(datos_parada1[2])
            lon1 = float(datos_parada1[3])
            lat2 = float(datos_parada2[2])
            lon2 = float(datos_parada2[3])

            d_lat = math.radians(lat2 - lat1)
            d_lon = math.radians(lon2 - lon1)
            a = math.sin(d_lat / 2) ** 2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(d_lon / 2) ** 2
            c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
            return radio_tierra * c
        except ValueError:
            print("Error: Las coordenadas no son números válidos.")
            return float("inf")

    def buscar_parada(self, lineas_y_paradas, nombre_parada):
        for linea in lineas_y_paradas:
            if len(linea) > 1 and linea[1] == nombre_parada:
                return linea
        return None

    def mostrar_resultados_finales(self, ruta_optimizada, lineas_y_paradas):
        if not ruta_optimizada:
            print("No se encontraron datos para mostrar.")
            return

        print("=== RUTA OPTIMIZADA ===")
        print("+---------------------+----------------------+----------------------+")
        print("| Nombre de Parada    | Latitud              | Longitud             |")
        print("+---------------------+----------------------+----------------------+")

        for parada in ruta_optimizada:
            datos_parada = self.buscar_parada(lineas_y_paradas, parada)
            if datos_parada and len(datos_parada) > 3:
                print(f"| {datos_parada[1]:<20} | {datos_parada[2]:<20} | {datos_parada[3]:<20} |")

        print("+---------------------+----------------------+----------------------+")